src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs

```diff
@@ -13,28 +16,102 @@ public WhenKeywordRecommender()
-            if (IsIdentifierInCasePattern(context.TargetToken.Parent))
+            expressionOrPattern = switchLabel.ChildNodes().FirstOrDefault();
+            if (expressionOrPattern == null) // Oh well. It must have been a default label.
             {
-                return true;
+                return false;
             }
 
-            return false;
+            // If the last token is missing, the expression is incomplete - possibly because of missing parentheses,
+            // but not necessarily. We don't want to offer 'when' in those cases. Here are some examples that illustrate this:
+            // case |
+            // case x.|
+            // case 1 + |
+            // case (1 + 1 |
+
+            // Also note that if there's a missing token inside the expression, that's fine and we do offer 'when':
+            // case (1 + ) |
+
+            var lastToken = expressionOrPattern.GetLastToken(includeZeroWidth: true);
+            if (lastToken.IsMissing)
+            {
+                return false;
+            }
+
+            if (lastToken == context.LeftToken && context.LeftToken != context.TargetToken &&
+                expressionOrPattern is DeclarationPatternSyntax declarationPattern)
+            {
+                // case constant w|
+
+                // The user is typing a new word (might be a partially written 'when' keyword), which causes this to be parsed
+                // as a declaration pattern. lastToken will be 'w' (LeftToken) as opposed to 'constant' (TargetToken).
+                // However we'd like to pretend that this is not the case and that we just a have single expression
+                // with 'constant' as if the new word didn't exist. So let's do that by adjusting our variables.
+
+                lastToken = context.TargetToken;
+                expressionOrPattern = declarationPattern.Type;
+            }
+
+            // Only offer if we're at the last token of the expression/pattern, not inside it.
+            return lastToken == context.TargetToken;
         }
 
-        private static bool IsIdentifierInCasePattern(SyntaxNode node)
+        private static bool IsTypeName(SyntaxNode expressionOrPattern, SemanticModel semanticModel,
+            CancellationToken cancellationToken)
         {
-            return node.IsKind(SyntaxKind.SingleVariableDesignation)
-                && node.IsParentKind(SyntaxKind.DeclarationPattern)
-                && node.Parent.IsParentKind(SyntaxKind.CasePatternSwitchLabel);
+            // Syntactically, everything works out. We're in a pretty good spot to show 'when' now.
+            // But let's not do it just yet... Consider these cases:
+            // case SyntaxNode |
+            // case SyntaxNode w|
+            // If what we have here is known to be a type, we don't want to clutter the variable name suggestion list
+            // with 'when' since we know that the resulting code would be semantically invalid.
```

Comment:
ignore my higher level request for commenting.  this comment here helps explain what's going on.